File: go-pubsub/publisher.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"

	"github.com/ably/ably-go/ably"
)

func main() {

	var (
		ABLY_KEY    = "xVLyHw.RnzM8g:iV7CnSzWDKG7HRFFOieUCYyDG6z1QYXMiqlk-RF0A6U"
		channelName = "test"
	)

	client, err := ably.NewREST(
		ably.WithKey(ABLY_KEY),
		ably.WithLogLevel(ably.LogWarning),
	)
	if err != nil {
		log.Fatal(err)
	}

	channel := client.Channels.Get(channelName)
	// channel.Attach(context.Background())
	// time.Sleep(time.Second)

	data, err := os.ReadFile(filepath.Join("../generated", "messages.json"))
	if err != nil {
		log.Fatal(err)
	}
	var messages []map[string]interface{}
	err = json.Unmarshal(data, &messages)
	if err != nil {
		log.Fatal(err)
	}

	ctx, cancel := context.WithCancel(context.Background())

	sentMessages := make([]map[string]interface{}, len(messages))
	for index := 0; index < len(messages); index++ {
		message := messages[index]
		err := channel.Publish(context.Background(), "greeting", message)
		if err != nil {
			fmt.Printf("Error publishing message %v\n", err)
		} else {
			fmt.Printf("published message %v\n", index)
			sentMessages[index] = message
		}
		if index == len(messages)-1 {
			cancel()
		}
	}

	<-ctx.Done()

	fmt.Printf("total messages published %v", len(sentMessages))
	serializedString, err := json.MarshalIndent(sentMessages, "", "    ")
	if err != nil {
		log.Fatal(err)
	}
	filepath := filepath.Join("../generated", "go_published.json")
	f, err := os.Create(filepath)
	if err != nil {
		log.Fatal(err)
	}
	_, er := f.WriteString(string(serializedString))
	if er != nil {
		log.Fatal(er)
	}
	defer f.Close()

}

File: js-pubsub/publisher.js
var Ably = require('ably');
const fs = require('fs');
const path = require('path');

const ABLY_KEY = "xVLyHw.RnzM8g:iV7CnSzWDKG7HRFFOieUCYyDG6z1QYXMiqlk-RF0A6U";
const publishedMessagesPath = '../generated' + path.sep + 'js_published.json';

var ablyClient = new Ably.Rest({ key: ABLY_KEY, logLevel: 'warn'});
var channel = ablyClient.channels.get('test');

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

const data = fs.readFileSync('../generated' + path.sep + 'messages.json', 'utf8');
const messagesToBePublished = JSON.parse(data)

publishMessages(messagesToBePublished).then(messagesPublished => {
    console.log("total messages sent "+ messagesPublished.length)
    fs.writeFileSync(publishedMessagesPath, JSON.stringify(messagesPublished, null, 2));
});

async function publishMessages(messagesToBePublished) {  
    var messagesPublished = [];
    var resolve = () => {};
    const promise = new Promise(res => resolve = res);
    for (let index = 0; index < messagesToBePublished.length; index++) {
        const message = messagesToBePublished[index];
        await sleep(10);
        channel.publish('greeting', message, function(err) {
            if(err) {
                console.log('publish failed with error ' + err);
            } else {
                messagesPublished.push(message);
                console.log('publish succeeded '+ index);
                if(messagesPublished.length == messagesToBePublished.length) {
                    resolve(messagesPublished);
                }
            }
        });
    }
    return promise;
}
  
